docs/blueprints/zombie-segments-cleaner.md

```diff
@@ -0,0 +1,55 @@
+# Zombie Segments Cleaner
+
+## Abstract
+
+This document describes a design for cleaning segments that are not accessible with standard operations like listing.
+
+## Background
+
+Currently with a multisegment object, we have several places where we can leave unfinished or broken objects:
+* if upload process will be stopped, because of an error or canceling, when one or more segments are already uploaded we can become zombie segments.
+* if delete process will be stopped, because of an error or canceling, when one or more segments are already deleted we can have an object with an incomplete number of segments.
+* in the past we had a bug where error during deleting segments from nodes was interrupting deleting segment from satellite, and we have segments available on satellite but not on storage node.
+* deleted bucket can leave all objects inaccessible ???
+* deleted project can leave all objects inaccessible ???
+
+We need a system for identifying objects with missing segments, and zombie segments that are not listed as part of objects, that cleans up those segments.
+
+In the long term, the metainfo refactor will fix this. We need a short term solution.
+
+## Design
+
+How detect zombie segment:
+* segment doesn't have corresponding last segment (`l`)
+* segment where index is greater than 0 but any previous segment is missing
+* last segment (`l`) where unencrypted number of segments is greater than 0 but rest of segments are missing or number of existing segments is different from stored value
+
+Different kind of bad segment is a case where all segments are available on satellite but were deleted from storage nodes.
+
+General idea is to register a new `metainfo.Observer` to verify all segment paths. During iteration, each segment path will be processed and assigned to a struct that will collect segments from the same object. When the last segment will be reached or iteration will be finished each object will be checked if it contains all segments (according to rules from the beginning of this part). If yes, then helper struct will be removed from memory. If not, then all related segments will be moved to delete.
```

Comment:
I wonder whether we should do this as a completely separate process?

Reply:
you are concerned that this will be too much overhead for satellite?

Reply:
Yeah and memory usage. Also we don't have to keep it running all the time.

Reply:
running this all the time has this advantage that we will be cleaning things reguraly. After first large execution next iterations will be much smaller from cpu/memory perspective. But maybe its not important right now.

Reply:
making separate command for this should work well for this too.

Reply:
If the observer gets the segments in sorted order, it should be able to be done in O(1) memory, right? No allocations really needed.

Reply:
If I am not wrong the structure of a segment path is like this:
```
<project-id>/<s0..N|l>/<bucket-name>/the/object/path/of/arbitrary/length
```
If the observer really gets the segments in sorter order, this would mean that it can clean up the collected data when finishing a project. The size of the allocated memory would be proportional to the project size.
```

Reply:
Ah, yep. That’s right.